From edc42b8b7e0c68e6277e5b4c5ddb392c8bd957e3 Mon Sep 17 00:00:00 2001
From: Kuntal Nayak <nkuntal@vmware.com>
Date: Sat, 26 Aug 2023 06:06:20 +0000
Subject: [PATCH] Fix CLI seg fault on missing -nonce

The upstream patch - https://sqlite.org/src/info/cd24178bbaad4a1d
is not directly applicable to the source that we are consuming because
source: https://www.sqlite.org/src/tarball/sqlite.tar.gz?r=release. Whilst
we are consuming: https://sqlite.org/2022/sqlite-autoconf-%{sourcever}.tar.gz. 

So the patch had to be adapted and file path is modified.

The bug in command-line argument parsing. Due to improper command-line
argument processing a NULL-pointer dereference will occur and the
"sqlite3" tool will fail to start up.

Detailed CVE and fix info: https://www.sqlite.org/forum/forumpost/19f55ef73b
---
 shell.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/shell.c b/shell.c
index d104768..cc8f7d9 100644
--- a/shell.c
+++ b/shell.c
@@ -23163,8 +23163,12 @@ int SQLITE_CDECL wmain(int argc, wchar_t **wargv){
     }else if( strcmp(z,"-bail")==0 ){
       bail_on_error = 1;
     }else if( strcmp(z,"-nonce")==0 ){
-      free(data.zNonce);
-      data.zNonce = strdup(argv[++i]);
+      if( data.zNonce ) free(data.zNonce);
+      if( i+1 < argc ) data.zNonce = strdup(argv[++i]);
+      else{
+        data.zNonce = 0;
+        break;
+      }
     }else if( strcmp(z,"-safe")==0 ){
       /* no-op - catch this on the second pass */
     }
-- 
2.39.0

